### PR TITLE
STEP 13 contacts notes entitlement

### PR DESCRIPTION
**Pull request by lpusok**
https://bitrise.atlassian.net/browse/STEP-13

Adding check whether an unsupported entitlement is used, that we can not add on the API.

- We check this before checking if App ID exist and is valid as as far we know,
these needed to be added in the provisioning profile generation phase.
So even if the user creates App ID manually, the step can not generate the correct profiles.
The extra entitlements require Apple approval, which I could not test,
so choose to fail early and make the user use the Certificates Installer Step.

Discussed with @istvankovacs-bitrise adding this logic to SyncBundleID -> Capabilities at https://github.com/bitrise-steplib/steps-ios-auto-provision-appstoreconnect/blob/8394b5a0f0f72b4a95728d6e4da3d7463144eeec/autoprovision/entitlements.go#L179, but decided that is not the same use case, as supposedly the user still can generate App ID manually, and the step will use it (may need the check this logic again as did not validate)

Carplay entitlement: https://developer.apple.com/documentation/carplay/requesting_the_carplay_entitlements
Contacts notes entitlement: https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_contacts_notes

Other entitlements for reference (not yet checked): https://developer.apple.com/documentation/bundleresources/entitlements

### CODE REVIEW

File: appstoreconnect/appstoreconnect.go [hunk 1]
```diff
@@ -102,7 +102,7 @@ func (c *Client) NewRequest(method, endpoint string, body interface{}) (*http.Re
 	endpoint = apiVersion + "/" + endpoint
 	u, err := c.BaseURL.Parse(endpoint)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("parsing endpoint failed, %v", err)
```
Comment (godrei) [inline appstoreconnect/appstoreconnect.go:105]: please use `:` to separate the context from the error, like `parsing endpoint failed: %s` 

https://golang.org/doc/effective_go.html#errors

File: appstoreconnect/capabilities.go [hunk 2]
```diff
@@ -78,6 +78,18 @@ var ServiceTypeByKey = map[string]CapabilityType{
 	"com.apple.developer.ubiquity-container-identifiers": Ignored,
 }
 
+// UnsupportedEntitlements contains entitlements not supported via the API and this step,
+// profile needs to be manually generated on Apple Developer Portal
+var UnsupportedEntitlements = []string{
```
Comment (godrei) [inline appstoreconnect/capabilities.go:83]: for ignored (does not appear in profile) entitlements, we have a special type `Ignore`, see: https://github.com/bitrise-steplib/steps-ios-auto-provision-appstoreconnect/pull/31/files#diff-9e4f54818286790433c9c643f7f188562f0ae208f007cc141c8d8d0b22361e80R16

what about unifying this solution and adding a new type: `unsupported` and handle both similarly?

File: main.go [hunk 3]
```diff
@@ -385,6 +385,10 @@ func main() {
 		log.Printf("- %s", id)
 	}
 
+	if err := autoprovision.ContainsUnsupportedEntitlement(entitlementsByBundleID, appstoreconnect.UnsupportedEntitlements); err != nil {
```
Comment (godrei) [inline main.go:388]: might it is better to suggest manually configuring the profile's unsupported entitlement, after that the step can work properly (as long as no new unsupported entitlements appear)

**Review by godrei — CHANGES_REQUESTED** [review]

File: main.go (hunk 3 above)
Comment (mateherber) [reply main.go:388]: Maybe we could list all not-supported entitlements as the error message

File: autoprovision/entitlements_test.go [hunk 4]
```diff
@@ -102,59 +102,73 @@ func TestICloudContainers(t *testing.T) {
...
-func TestContainsUnsupportedEntitlement(t *testing.T) {
+func TestCanGenerateProfileWithEntitlements(t *testing.T) {
 	tests := []struct {
-		name                    string
-		entitlementsByBundleID  map[string]serialized.Object
-		unsupportedEntitlements []string
-		wantErr                 bool
+		name                   string
+		entitlementsByBundleID map[string]serialized.Object
+		want                   bool
```
Comment (godrei) [inline autoprovision/entitlements_test.go:109]: pls use more descriptive properties, like `wantOk`, `wantEntitlement `, `wantBundleID `

File: autoprovision/entitlements_test.go [hunk 5]
```diff
@@ -102,59 +102,73 @@ func TestICloudContainers(t *testing.T) {
...
+			want:  true,
+			want1: "",
+			want2: "",
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if err := autoprovision.ContainsUnsupportedEntitlement(tt.entitlementsByBundleID, tt.unsupportedEntitlements); (err != nil) != tt.wantErr {
-				t.Errorf("ContainsUnsupportedEntitlement() error = %v, wantErr %v", err, tt.wantErr)
+			got, got1, got2 := autoprovision.CanGenerateProfileWithEntitlements(tt.entitlementsByBundleID)
```
Comment (godrei) [inline autoprovision/entitlements_test.go:163]: pls use more descriptive vars, like `gotOk`, `gotEntitlement`, `gotBundleID`

File: main.go [hunk 6]
```diff
@@ -385,8 +385,9 @@ func main() {
 		log.Printf("- %s", id)
 	}
 
-	if err := autoprovision.ContainsUnsupportedEntitlement(entitlementsByBundleID, appstoreconnect.UnsupportedEntitlements); err != nil {
-		failf("Error: %v. Generate provisioning profile manually on Apple Developer Portal and use the Certificate and profile installer Step.", err)
+	if ok, entitlement, bundleID := autoprovision.CanGenerateProfileWithEntitlements(entitlementsByBundleID); !ok {
+		log.Errorf("Can not create profile with unsupported entitlement (%s) for the bundle ID %s, due to API limitations.", entitlement, bundleID)
```
Comment (godrei) [inline main.go:389]: emphasize that it is not our service limitation: `due to API limitations` -> `due to App Store Connect API limitations`

**Review by godrei — CHANGES_REQUESTED** [review]
pls check my notes

File: main.go (hunk 3 above)
Comment (lpusok) [reply main.go]: Discussed that we can not support this without sample app.

File: appstoreconnect/appstoreconnect.go (hunk 1 above)
Comment (lpusok) [reply appstoreconnect/appstoreconnect.go]: Done.

File: autoprovision/entitlements_test.go (hunk 4 above)
Comment (lpusok) [reply autoprovision/entitlements_test.go]: done

File: main.go (hunk 6 above)
Comment (lpusok) [reply main.go]: done

**Review by godrei — APPROVED** [review]

**Comment by petermnt** [comment]
@lpusok  There is actually an undocumented way to create a provisioning profile with this special entitlement through the Appstore Connect api.
Fastlane has this implemented and this works for us. See https://github.com/fastlane/fastlane/blob/2624a24407dee94f67d7e39b5651f9fdc41dcb00/spaceship/lib/spaceship/connect_api/models/profile.rb#L83

The tricky thing is that this `templateName` is a very specific string that might not even be the same for every developer.
For us this is 
* CarPlay Audio App(Distribution)
* CarPlay Audio App(Distribution) (Adhoc)
* CarPlay Audio App(Develkopment)

Passing these to the API will create the correct provisioning profile.

Maybe `templateName` could be added as a configurable field in the bitrise step for people who need this?

**Comment by lpusok** [comment]
@petermnt Checking, was not aware of this. How do you specifically look up the template name? In case I implement this, can you maybe help to validate that it works correctly?

**Comment by petermnt** [comment]
@lpusok  The `templateName` works for creating new profiles.
The only way I know to look up the template name is to look manually in the developer portal.

There is a dropdown where you choose the special entitlement.
There can only be 1 per profile.

It looks like this:

CleanShot 2020-12-09 at 19 10 53@2x

There is another thing to pay attention to, which is that (for us) in Adhoc configuration the visible value in the dropdown does not match the actual value:

```html
<select id="template" name="template"><option value="Standard">Default</option><option value="CarPlay Audio App(Distribution) (Adhoc)" selected="">CarPlay Audio App(Distribution)</option></select>
<option value="Standard">Default</option>
<option value="CarPlay Audio App(Distribution) (Adhoc)" selected="">CarPlay Audio App(Distribution)</option>
<select id="template" name="template"><option value="Standard">Default</option><option value="CarPlay Audio App(Distribution) (Adhoc)" selected="">CarPlay Audio App(Distribution)</option></select>
```

Passing the `value` `CarPlay Audio App(Distribution) (Adhoc)` to `templateName` creates the profile with this dropdown properly select.

**Pull request merged**